From 1151dd2b0f32424435c47685a33933dc0a7d4c35 Mon Sep 17 00:00:00 2001
From: Jason Lowe <jlowe@yahoo-inc.com>
Date: Tue, 28 Mar 2017 09:33:26 -0500
Subject: [PATCH 2352/2748] YARN-6359.
 TestRM#testApplicationKillAtAcceptedState fails
 rarely due to race condition. Contributed by
 Robert Kanter

(cherry picked from commit bea8f8190f036375981ef4ceda409f9199730740)

Conflicts:
	hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/TestRM.java

Change-Id: I3d6aefe280adbc133af90727db90a8e5879489b1
---
 .../hadoop/yarn/server/resourcemanager/TestRM.java |   16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/TestRM.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/TestRM.java
index b04b173..029b8bb 100644
--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/TestRM.java
+++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/test/java/org/apache/hadoop/yarn/server/resourcemanager/TestRM.java
@@ -18,6 +18,8 @@
 
 package org.apache.hadoop.yarn.server.resourcemanager;
 
+import com.google.common.base.Supplier;
+import org.apache.hadoop.test.GenericTestUtils;
 import org.junit.Before;
 import static org.mockito.Matchers.argThat;
 import static org.mockito.Mockito.doNothing;
@@ -597,9 +599,9 @@ protected Dispatcher createDispatcher() {
     };
 
     // test metrics
-    QueueMetrics metrics = rm.getResourceScheduler().getRootQueueMetrics();
-    int appsKilled = metrics.getAppsKilled();
-    int appsSubmitted = metrics.getAppsSubmitted();
+    final QueueMetrics metrics = rm.getResourceScheduler().getRootQueueMetrics();
+    final int appsKilled = metrics.getAppsKilled();
+    final int appsSubmitted = metrics.getAppsSubmitted();
 
     rm.start();
     
@@ -637,7 +639,13 @@ protected Dispatcher createDispatcher() {
     rm.waitForState(application.getApplicationId(), RMAppState.KILLED);
 
     // test metrics
-    metrics = rm.getResourceScheduler().getRootQueueMetrics();
+    GenericTestUtils.waitFor(new Supplier<Boolean>() {
+      @Override
+      public Boolean get() {
+        return appsKilled + 1 == metrics.getAppsKilled()
+            && appsSubmitted + 1 == metrics.getAppsSubmitted();
+      }
+    }, 100, 10000);
     Assert.assertEquals(appsKilled + 1, metrics.getAppsKilled());
     Assert.assertEquals(appsSubmitted + 1, metrics.getAppsSubmitted());
   }
-- 
1.7.9.5

